File: javascript/testLevy.js
const ConstrainedGA = require('./constrainedGA');

// Section 3.1: LEVY Problem Setup
const levyObjective = (x) => -x[0] - x[1];
const levyConstraint = (x) => {
    const a = 2.0;
    const b = 0.25;
    const term1 = (Math.pow(x[0] - 1, 2) + (x[1] - 1)) * ((1 / (2 * Math.pow(a, 2))) - (1 / (2 * Math.pow(b, 2))));
    const term2 = (x[0] - 1) * (x[1] - 1) * ((1 / Math.pow(a, 2)) - (1 / Math.pow(b, 2)));
    return -(term1 + term2 - 1);
};
const levyBounds = [[0.0, 1.0], [0.0, 1.0]];

const runsCount = 100;
const executionValues = [];
const executionFevals = [];

console.log(`Starting validation: Running LEVY optimization ${runsCount} times...`);

for (let run = 1; run <= runsCount; run++) {
    const solver = new ConstrainedGA({
        objFunc: levyObjective,
        inequalityConstraints: [levyConstraint],
        bounds: levyBounds,
        K: 200,
        iterMax: 200,
        lam: 1000.0
    });

    const runResult = solver.solve();
    executionValues.push(runResult.value);
    executionFevals.push(runResult.fevalsSpent);
    
    if (run % 20 === 0) {
        console.log(`  > Completed ${run}/${runsCount} runs...`);
    }
}

// Compute statistics
const avgFevals = executionFevals.reduce((a, b) => a + b, 0) / runsCount;
const bestValue = Math.min(...executionValues);
const meanValue = executionValues.reduce((a, b) => a + b, 0) / runsCount;
const variance = executionValues.reduce((sum, val) => sum + Math.pow(val - meanValue, 2), 0) / runsCount;
const stdDeviation = Math.sqrt(variance);

// Benchmark Table Output
console.log("\n=======================================================");
console.log("            LEVY PROBLEM BENCHMARK RESULTS             ");
console.log("=======================================================");
console.log(`Metric       |  Paper Table 2 Data  | Your Local Run JS`);
console.log("-------------------------------------------------------");
console.log(`FEVALS       |  4572               | ${Math.round(avgFevals)}`);
console.log(`BEST         |  -1.8730            | ${bestValue.toFixed(4)}`);
console.log(`MEAN         |  -1.8730            | ${meanValue.toFixed(4)}`);
console.log(`STD          |  4.82e-6            | ${stdDeviation.toExponential(2)}`);
console.log("=======================================================");

File: matlab/testLevy.m
% LEVY Problem Test
% Section 3.1: LEVY Problem Setup

clear; clc;

% Define objective function
levyObjective = @(x) -x(1) - x(2);

% Define constraint
levyConstraint = @(x) levyConstraintFunc(x);

% Variable bounds
levyBounds = [0.0 1.0; 0.0 1.0];

% Number of runs
runsCount = 100;
executionValues = zeros(runsCount, 1);
executionFevals = zeros(runsCount, 1);

fprintf('Starting validation: Running LEVY optimization %d times...\n', runsCount);

for run = 1:runsCount
    solver = ConstrainedGA('ObjectiveFunction', levyObjective, ...
                           'InequalityConstraints', {levyConstraint}, ...
                           'Bounds', levyBounds, ...
                           'PopulationSize', 200, ...
                           'MaxGenerations', 200, ...
                           'PenaltyMultiplier', 1000.0);
    
    runResult = solver.solve();
    executionValues(run) = runResult.value;
    executionFevals(run) = runResult.fevalsSpent;
    
    if mod(run, 20) == 0
        fprintf('  > Completed %d/%d runs...\n', run, runsCount);
    end
end

% Compute statistics
avgFevals = mean(executionFevals);
bestValue = min(executionValues);
meanValue = mean(executionValues);
stdDeviation = std(executionValues);

% Display results
fprintf('\n=======================================================\n');
fprintf('            LEVY PROBLEM BENCHMARK RESULTS             \n');
fprintf('=======================================================\n');
fprintf('Metric       |  Paper Table 2 Data  | MATLAB Run\n');
fprintf('-------------------------------------------------------\n');
fprintf('FEVALS       |  4572                | %d\n', round(avgFevals));
fprintf('BEST         |  -1.8730             | %.4f\n', bestValue);
fprintf('MEAN         |  -1.8730             | %.4f\n', meanValue);
fprintf('STD          |  4.82e-6             | %.2e\n', stdDeviation);
fprintf('=======================================================\n');

% Helper function for LEVY constraint
function val = levyConstraintFunc(x)
    a = 2.0;
    b = 0.25;
    term1 = ((x(1) - 1)^2 + (x(2) - 1)) * ((1 / (2 * a^2)) - (1 / (2 * b^2)));
    term2 = (x(1) - 1) * (x(2) - 1) * ((1 / a^2) - (1 / b^2));
    val = -(term1 + term2 - 1);
end
